Atomic/util.py

```diff
@@ -802,102 +801,79 @@ def load_scan_result_file(file_name):
-    def _read_install_data(file_handle):
-        try:
-            return json.loads(file_handle.read())
-        except ValueError:
-            return {}
-
-    @classmethod
-    def _write_install_data(cls, new_data):
-        install_data = cls._read_install_data(cls.install_file_handle)
-        for x in new_data:
-            install_data[x] = new_data[x]
-        temp_file = tempfile.NamedTemporaryFile(mode='w', delete=False)
-        json.dump(install_data, temp_file)
-        temp_file.close()
-        shutil.move(temp_file.name, ATOMIC_INSTALL_JSON)
 
     @classmethod
-    @file_lock
     def read_install_data(cls):
-        if os.path.exists(ATOMIC_INSTALL_JSON):
-            read_data = cls._read_install_data(cls.install_file_handle)
-            return read_data
-        return {}
+        with file_lock(ATOMIC_INSTALL_JSON):
+            try:
+                with open(ATOMIC_INSTALL_JSON, 'r') as f:
+                    # Backwards compatibilty - we previously created an empty file explicitly;
+                    # see https://github.com/projectatomic/atomic/pull/966
+                    if os.fstat(f.fileno()).st_size == 0:
+                        return {}
+                    return json.load(f)
+            except IOError as e:
+                if e.errno == errno.ENOENT:
+                    return {}
+                raise e
 
     @classmethod
-    @file_lock
     def write_install_data(cls, new_data):
-        cls._write_install_data(new_data)
+        install_data = cls.read_install_data()
+        with file_lock(ATOMIC_INSTALL_JSON):
+            for x in new_data:
+                install_data[x] = new_data[x]
+                temp_file = tempfile.NamedTemporaryFile(mode='w', delete=False)
+                json.dump(install_data, temp_file)
+                temp_file.close()
+                if not os.path.exists(ATOMIC_VAR_LIB):
+                    os.makedirs(ATOMIC_VAR_LIB)
+                shutil.move(temp_file.name, ATOMIC_INSTALL_JSON)
 
     @classmethod
     def get_install_name_by_id(cls, iid, install_data=None):
         if not install_data:
-            install_data = cls._read_install_data(cls.install_file_handle)
+            install_data = cls.read_install_data()
         for installed_image in install_data:
             if install_data[installed_image]['id'] == iid:
                 return installed_image
         raise ValueError("Unable to find {} in installed image data ({}). Re-run command with -i to ignore".format(id, ATOMIC_INSTALL_JSON))
 
     @classmethod
-    @file_lock
     def delete_by_id(cls, iid, ignore=False):
-        install_data = cls._read_install_data(cls.install_file_handle)
+        install_data = cls.read_install_data()
         try:
             id_key = InstallData.get_install_name_by_id(iid, install_data=install_data)
         except ValueError as e:
             if not ignore:
                 raise ValueError(str(e))
             return
         del install_data[id_key]
```

Comment:
Wait, this won't take effect, right? The `write` methods (both old and new) are merging entries, so can deletes actually occur?